### PR TITLE
Fix a bug in sample!

### LINKED ISSUE #196: Problems with the `sample!` function
The sample function does not do what it claims.

1. `sample!` with `replace` is not tested.
2. `sample!` with `replace` does not work because:
```julia
    for (index, id) in enumerate(newids) # add new agents while adjusting id
        model.agents[index] = deepcopy(model[id])
        model.agents[index].id = index
    end
```
thus an agent that has been selected twice is only added once in the model.

3. The design of the `sample!` function has bad performance. It removes and re-adds everything from scratch, but I don't understand why. Why not just kill the agents that were not selected, and add any additional extra ones (e.g. add the extra duplicates if `resample` , or just add the extra created agents if `n > nagents(model)` ?

### PR DESCRIPTION
**Pull request by kavir1698**
Closes #196

### CODE REVIEW

File: src/simulations/sample.jl [hunk 1]
```diff
@@ -13,51 +13,34 @@ this agent will be chosen in the new sampling.
...
-        model.agents[index].id = index
-    end
-    # kill extra agents
-    if n < nagents(model)
-        genocide!(model, n)
+      noccurances = count(x->x==id, newids)
+      for t in 2:noccurances
+        newagent = deepcopy(model.agents[id])
+        newagent.id = nextid
+        add_agent_pos!(newagent, model)
```
Comment (Libbum) [inline src/simulations/sample.jl:41]: Tests are failing at this point. The test `Agent2` has no position.

Comment (Libbum) [reply src/simulations/sample.jl:41]: Could add a gate here with something like `hasfield(typeof(newagent),:pos)`?

**Comment by Libbum** [comment]
Whilst we're fixing bugs in `sample!`, we should also add tests and logic to cover the case when a user attempts to sample from a pool of zero agents:

```julia
julia> model2 = ABM(Agent3, GridSpace((10, 10)))
AgentBasedModel with 0 agents of type Agent3
 space: GridSpace with 100 nodes and 180 edges
 scheduler: fastest

julia> sample!(model2, 10, :weight)
ERROR: MethodError: no method matching zero(::Type{Any})
```

**Comment by Datseris** [comment]
We should add tests for `sample!` with replace in this pr as well.

**Comment by kavir1698** [comment]
> nagents(model) > 0 || throw(ArgumentError("Model must have at least 1 agent to sample from."))
I think it's better to return nothing

**Comment by Libbum** [comment]
Huh. We have uncovered another bug. There is no ability to add an instantiated agent without a space.

**Comment by kavir1698** [comment]
> We should add tests for sample! with replace in this pr as well.

sampling with replacement is the default behavior. So current tests cover it.

**Comment by kavir1698** [comment]
> Huh. We have uncovered another bug. There is no ability to add an instantiated agent without a space.

Are you talking about a model with `Agent2`? I don't receive an error. I am not on master, though.

**Comment by Datseris** [comment]
`hasfield(A, :pos)` can't be used for mixed agent models.

**Comment by Libbum** [comment]
> I don't receive an error.

That's odd. You should! I do when I check out this branch locally, [so does Travis](https://travis-ci.com/github/JuliaDynamics/Agents.jl/jobs/315941841#L359-L361).

**Comment by kavir1698** [comment]
> That's odd. You should! I do when I check out this branch locally, so does Travis.

Maybe it's just because I am using the stable version. I will push a commit with a new `add_agent!` defined.

**Comment by Datseris** [comment]
I've fixed the missing method. But @kavir1698 the tests still fail both locally and oline for other reason, I think you need to have a second look at the new `sample!` (some Keys are wrong).

**Comment by Libbum** [comment]
I think this might actually be a test design flaw, more than the `sample!` implementation.

line 153 in api_tests.jl
```julia
julia> model.agents[1].weight = 2
ERROR: KeyError: key 1 not found

julia> model.agents
Dict{Int64,Agent2} with 12 entries:
  2  => Agent2(2, 3.0)
  11 => Agent2(11, 0.573621)
  21 => Agent2(21, 2.13182)
  7  => Agent2(7, 1.62655)
  26 => Agent2(26, 3.955)
  25 => Agent2(25, 3.955)
  10 => Agent2(10, 0.691542)
  22 => Agent2(22, 2.13182)
  24 => Agent2(24, 3.955)
  4  => Agent2(4, 0.987061)
  23 => Agent2(23, 3.955)
  15 => Agent2(15, 3.955)
```

Since sample! is not expected to re-align ids, I'll change these tests to take random agents instead of specific ones.

**Comment by kavir1698** [comment]
> Don't even make sense? Should we just drop these?

These were to catch a bug that was caused by not using `deepcopy`. In their current form and the new code that changes all agent ids, they don't make sense anymore.
